File: server/app/models/pieces.scala
package models

sealed trait Piece {
  val player: Player.Value
}
sealed trait PromotablePiece extends Piece {
  val promoted: Boolean
}
case class King(player: Player.Value) extends Piece
case class Rook(player: Player.Value, promoted: Boolean = false) extends PromotablePiece
case class Bishop(player: Player.Value, promoted: Boolean = false) extends PromotablePiece
case class Gold(player: Player.Value) extends Piece
case class Silver(player: Player.Value, promoted: Boolean = false) extends PromotablePiece
case class Knight(player: Player.Value, promoted: Boolean = false) extends PromotablePiece
case class Lance(player: Player.Value, promoted: Boolean = false) extends PromotablePiece
case class Pawn(player: Player.Value, promoted: Boolean = false) extends PromotablePiece

object Piece {
  def unapply(p: Piece): Option[Player.Value] = Some(p.player)

  def pieceName(p: Piece): String = p match {
    case Pawn(_, _) => "pawn"
    case Bishop(_, _) => "bishop"
    case Rook(_, _) => "rook"
    case Lance(_, _) => "lance"
    case Knight(_, _) => "knight"
    case Silver(_, _) => "silver"
    case Gold(_) => "gold"
    case King(_) => "king"
  }

  def unpromotedPiece(p: Piece): Piece = p match {
    case Pawn(player, true) => Pawn(player)
    case Rook(player, true) => Rook(player)
    case Bishop(player, true) => Bishop(player)
    case Lance(player, true) => Lance(player)
    case Knight(player, true) => Knight(player)
    case Silver(player, true) => Silver(player)
    case _ => p
  }

  def promotedPiece(p: Piece): Piece = p match {
    case Pawn(player, _) => Pawn(player, promoted = true)
    case Rook(player, _) => Rook(player, promoted = true)
    case Bishop(player, _) => Bishop(player, promoted = true)
    case Lance(player, _) => Lance(player, promoted = true)
    case Knight(player, _) => Knight(player, promoted = true)
    case Silver(player, _) => Silver(player, promoted = true)
    case _ => p
  }
}

object PromotablePiece {
  def unapply(p: PromotablePiece): Option[(Player.Value, Boolean)] = Some(p.player, p.promoted)
}

File: client/src/app/models/pieces.ts
import { Player } from './player';

export class NonPromotablePiece {
    constructor(readonly player: Player) {};
}

export class PromotablePiece {
    constructor(readonly player: Player, readonly promoted: boolean) {};
}

export class King extends NonPromotablePiece {
    readonly type = 'king';
    constructor(player: Player) {
        super(player);
    }
}

export class Gold extends NonPromotablePiece {
    readonly type = 'gold';
    constructor(player: Player) {
        super(player);
    }
}

export class Silver extends PromotablePiece {
    readonly type = 'silver';
    constructor(player: Player, promoted: boolean) {
        super(player, promoted);
    }
}

export class Knight extends PromotablePiece {
    readonly type = 'knight';
    constructor(player: Player, promoted: boolean) {
        super(player, promoted);
    }
}

export class Lance extends PromotablePiece {
    readonly type = 'lance';
    constructor(player: Player, promoted: boolean) {
        super(player, promoted);
    }
}

export class Rook extends PromotablePiece {
    readonly type = 'rook';
    constructor(player: Player, promoted: boolean) {
        super(player, promoted);
    }
}

export class Bishop extends PromotablePiece {
    readonly type = 'bishop';
    constructor(player: Player, promoted: boolean) {
        super(player, promoted);
    }
}

export class Pawn extends PromotablePiece {
    readonly type = 'pawn';
    constructor(player: Player, promoted: boolean) {
        super(player, promoted);
    }
}

export type Piece = King | Gold | Silver | Knight | Lance | Rook | Bishop | Pawn;
